redcap_bridge/project_validation.py

```diff
@@ -0,0 +1,194 @@
+import pathlib
+import shutil
+import json
+import redcap
+import zipfile
+import pandas as pd
+from redcap_bridge.utils import map_header_json_to_csv
+
+index_column_header = 'Variable / Field Name'
+
+
+def validate_project_against_template_parts(project, *templates):
+    """
+    Validate a built project csv
+    Args:
+        project: (str, buffer)
+            Filepath of the csv file or csv buffer of the built project
+        templates: (str, list)
+            List of file paths of the template part csvs.
+
+    Returns:
+        (bool): True if the validation was successful
+    """
+
+    df_project = pd.read_csv(project)
+    # unify column names to conform to csv style
+    df_project = df_project.rename(columns=map_header_json_to_csv)
+    df_project.index = df_project[index_column_header]
+    dfs_templates = []
+
+    if not templates:
+        raise ValueError('No template_parts to validate against were specified.')
+
+    for template in templates:
+        df_template = pd.read_csv(template)
+        df_template.index = df_template[index_column_header]
+        dfs_templates.append(df_template)
+
+    # compare content of template_parts and project
+    for template_df in dfs_templates:
+        if not all(template_df.columns == df_project.columns):
+            raise ValueError(f'Incompatible columns in project '
+                             f'({project.columns}) and template '
+                             f'({template_df.columns})')
+
+        for i in template_df.index:
+            if i not in df_project.index:
+                raise ValueError(f'Row {i} is missing in project csv')
+
+            # compare entries of the row and exclude `na` entries
+            na_values = template_df.loc[i].isna()
+            equal_entries = df_project.loc[i] == template_df.loc[i]
+            if not (equal_entries | na_values).all():
+                raise ValueError(f'Row {i} is differs between project csv and '
+                                 f'template')
+
+    print('Validation successful')
+    return True
+
+def validate_record_against_template(template, record):
+    """
+    Validate a RedCap record against a template instrument
+
+    Args:
+        template: (dataframe) template structure of an instrument
+        record: (dataframe) data of a single record
+
+    Returns:
+        True
+
+    Raises:
+        ValueError in case of failing validation
```

Comment:
```suggestion
   Parameters
   ----------
   template: dataframe
       template structure of an instrument
   record: dataframe
       data of a single record

    Returns
    ----------
     True

    Raises
    ----------
    ValueError in case of failing validation
```

Reply:
Always returning true ? Seems a bit strange

Reply:
Yes, I was thinking about the case where you might at a later point not use valueerrors, but a bool instead. But maybe this is planned too far for now...